File: cpp/src/common/visualize.cpp
#include "edgeai/common/visualize.hpp"

#include <algorithm>
#include <cmath>
#include <iomanip>
#include <sstream>
#include <stdexcept>

#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

namespace edgeai::common {

cv::Scalar class_color(int class_id) {
    if (class_id < 0) {
        throw std::runtime_error("class_id must not be negative");
    }
    return {
        static_cast<double>((37 * class_id + 67) % 192 + 32),
        static_cast<double>((17 * class_id + 149) % 192 + 32),
        static_cast<double>((29 * class_id + 101) % 192 + 32),
    };
}

cv::Mat draw_detections(const cv::Mat& image, const std::vector<Detection>& detections) {
    if (image.empty() || image.dims != 2 || image.type() != CV_8UC3) {
        throw std::runtime_error("visualization expects a nonempty CV_8UC3 BGR image");
    }
    cv::Mat output = image.clone();
    for (const auto& detection : detections) {
        if (detection.class_id < 0 || detection.class_name.empty() ||
            !std::isfinite(detection.confidence) || detection.confidence < 0.0F ||
            detection.confidence > 1.0F) {
            throw std::runtime_error("detection metadata is invalid for visualization");
        }
        const int left = std::clamp(
            static_cast<int>(std::lround(detection.box_xyxy_source.x1)), 0, output.cols - 1
        );
        const int top = std::clamp(
            static_cast<int>(std::lround(detection.box_xyxy_source.y1)), 0, output.rows - 1
        );
        const int right = std::clamp(
            static_cast<int>(std::lround(detection.box_xyxy_source.x2)), 0, output.cols - 1
        );
        const int bottom = std::clamp(
            static_cast<int>(std::lround(detection.box_xyxy_source.y2)), 0, output.rows - 1
        );
        if (right <= left || bottom <= top) {
            throw std::runtime_error("detection box is invalid after raster clipping");
        }

        const cv::Scalar color = class_color(detection.class_id);
        std::ostringstream label_stream;
        label_stream << detection.class_name << ' ' << std::fixed << std::setprecision(3)
                     << detection.confidence;
        const std::string label = label_stream.str();
        cv::rectangle(output, {left, top}, {right, bottom}, color, 2, cv::LINE_8);
        int baseline = 0;
        const cv::Size text_size =
            cv::getTextSize(label, cv::FONT_HERSHEY_SIMPLEX, 0.6, 1, &baseline);
        const int label_top = std::max(0, top - text_size.height - baseline - 4);
        const int label_right = std::min(output.cols - 1, left + text_size.width + 4);
        cv::rectangle(output, {left, label_top}, {label_right, top}, color, cv::FILLED, cv::LINE_8);
        cv::putText(
            output,
            label,
            {left + 2, std::max(text_size.height + 1, top - baseline - 2)},
            cv::FONT_HERSHEY_SIMPLEX,
            0.6,
            {255, 255, 255},
            1,
            cv::LINE_AA
        );
    }
    return output;
}

void save_image(const std::filesystem::path& path, const cv::Mat& image) {
    if (image.empty()) {
        throw std::runtime_error("cannot save an empty image");
    }
    if (!path.parent_path().empty()) {
        std::filesystem::create_directories(path.parent_path());
    }
    if (!cv::imwrite(path.string(), image)) {
        throw std::runtime_error("OpenCV failed to write image: " + path.string());
    }
    if (!std::filesystem::is_regular_file(path) || std::filesystem::file_size(path) == 0U) {
        throw std::runtime_error("output image is missing or empty: " + path.string());
    }
}

}  // namespace edgeai::common

File: python/edgeai_benchmark/visualize.py
"""Deterministic OpenCV visualization for structured detections."""

from __future__ import annotations

from pathlib import Path
from typing import Any

import cv2
import numpy as np


class VisualizationError(RuntimeError):
    """Raised when detections cannot be rendered or saved."""


def class_color(class_id: int) -> tuple[int, int, int]:
    """Return a stable high-contrast BGR color for a class ID."""
    return (
        int((37 * class_id + 67) % 192 + 32),
        int((17 * class_id + 149) % 192 + 32),
        int((29 * class_id + 101) % 192 + 32),
    )


def draw_detections(
    image: np.ndarray,
    detections: list[dict[str, Any]],
) -> np.ndarray:
    """Draw on a copy so preprocessing source pixels are never mutated."""
    if image.ndim != 3 or image.shape[2] != 3 or image.dtype != np.uint8:
        raise VisualizationError(
            f"expected uint8 BGR image, observed shape={image.shape} dtype={image.dtype}"
        )
    output = image.copy()
    height, width = output.shape[:2]
    for detection in detections:
        x1, y1, x2, y2 = detection["box_xyxy_source"]
        left = min(max(int(round(x1)), 0), width - 1)
        top = min(max(int(round(y1)), 0), height - 1)
        right = min(max(int(round(x2)), 0), width - 1)
        bottom = min(max(int(round(y2)), 0), height - 1)
        if right <= left or bottom <= top:
            raise VisualizationError(f"invalid detection box after raster clipping: {detection}")
        color = class_color(int(detection["class_id"]))
        label = f"{detection['class_name']} {float(detection['confidence']):.3f}"
        cv2.rectangle(output, (left, top), (right, bottom), color, 2, cv2.LINE_8)
        (text_width, text_height), baseline = cv2.getTextSize(
            label,
            cv2.FONT_HERSHEY_SIMPLEX,
            0.6,
            1,
        )
        label_top = max(0, top - text_height - baseline - 4)
        label_right = min(width - 1, left + text_width + 4)
        cv2.rectangle(
            output,
            (left, label_top),
            (label_right, top),
            color,
            cv2.FILLED,
            cv2.LINE_8,
        )
        cv2.putText(
            output,
            label,
            (left + 2, max(text_height + 1, top - baseline - 2)),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.6,
            (255, 255, 255),
            1,
            cv2.LINE_AA,
        )
    return output


def save_png(path: Path, image: np.ndarray) -> None:
    """Write a PNG and reject silent OpenCV write failures."""
    path.parent.mkdir(parents=True, exist_ok=True)
    if not cv2.imwrite(str(path), image):
        raise VisualizationError(f"OpenCV failed to write output image: {path}")
    if not path.is_file() or path.stat().st_size <= 0:
        raise VisualizationError(f"output image is missing or empty after write: {path}")
